File: ios/Classes/FLTFFconfigPlugin.swift
//
//  FLTFFconfigPlugin.swift
//  ffmpeg_flutter
//
//  Created by nullptrX on 2021/4/6.
//

import Flutter
import UIKit

public class FLTFFconfigPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    
    private var events: FlutterEventSink?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "nullptrx.github.io/ffconfig", binaryMessenger: registrar.messenger())
        let instance = FLTFFconfigPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        let eventChannel = FlutterEventChannel(name: "nullptrx.github.io/ffconfig_event", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)
    }

    /// Uses `FlutterEventSink` to fire event.
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
//        examples:
//        events(true) // any generic type or more compex dictionary of [String:Any]
//        events(FlutterError(code: "ERROR_CODE",
//                            message: "Detailed message",
//                            details: nil)) // in case of errors
//        events(FlutterEndOfEventStream) // when stream is over
        self.events = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.events = nil
        return nil
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        result("iOS " + UIDevice.current.systemVersion)
    }

}

File: android/src/main/kotlin/io/github/nullptrx/ffmpegflutter/FLTFFconfigPlugin.kt
package io.github.nullptrx.ffmpegflutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class FLTFFconfigPlugin : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

  companion object {
    val kChannelName = "nullptrx.github.io/ffconfig"
    val kEventChannelName = "nullptrx.github.io/ffconfig_event"
  }

  private lateinit var methodChannel: MethodChannel
  private var events: EventChannel.EventSink? = null

  fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel = MethodChannel(binding.binaryMessenger, kChannelName)
    methodChannel.setMethodCallHandler(this)

    val eventChannel = EventChannel(binding.binaryMessenger, kEventChannelName)
    eventChannel.setStreamHandler(this)
  }

  fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    this.events = events
  }

  override fun onCancel(arguments: Any?) {
    this.events = null
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {


  }

}